odm2api/ODM2/services/readService.py

```diff
@@ -686,23 +687,109 @@ def getResults(self, ids=None, type=None, uuids=None, actionid=None, simulationi
+
+        Examples:
+            >>> READ = ReadODM2(SESSION_FACTORY)
+            >>> READ.getDataSets(ids=[39, 40])
+            >>> READ.getDataSets(codes=['HOME', 'FIELD'])
+            >>> READ.getDataSets(uuids=['a6f114f1-5416-4606-ae10-23be32dbc202',
+            ...                                 '5396fdf3-ceb3-46b6-aaf9-454a37278bb4'])
+            >>> READ.getDataSets(dstype='singleTimeSeries')
+
         """
         q = self._session.query(DataSets)
+        if ids:
+            q = q.filter(DataSets.DataSetID.in_(ids))
         if codes:
             q = q.filter(DataSets.DataSetCode.in_(codes))
         if uuids:
             q.filter(DataSets.DataSetUUID.in_(uuids))
+        if dstype:
+            q = q.filter(DataSets.DataSetTypeCV == dstype)
         try:
             return q.all()
         except Exception as e:
             print('Error running Query {}'.format(e))
             return None
 
+            # Datasets
+
+    def getDataSetsResults(self, ids= None, codes=None, uuids=None, dstype=None):
+        """
+        Retrieve a detailed list of Datasets along with detailed metadata about the datasets
+                and the results contained within them
+
+        **Must specify either DataSetID OR DataSetUUID OR DataSetCode)**
+        Args:
+            ids (list, optional): List of DataSetsIDs.
+            codes (list, optional): List of DataSet Codes.
+            uuids (list, optional): List of Dataset UUIDs string.
+            dstype (str, optional): Type of Dataset from
+                `controlled vocabulary name <http://vocabulary.odm2.org/datasettype/>`_.
+
+
+        Returns:
+            list: List of DataSetsResults Objects
+
+        Examples:
+            >>> READ = ReadODM2(SESSION_FACTORY)
+            >>> READ.getDataSetsResults(ids=[39, 40])
+            >>> READ.getDataSetsResults(codes=['HOME', 'FIELD'])
+            >>> READ.getDataSetsResults(uuids=['a6f114f1-5416-4606-ae10-23be32dbc202',
+            ...                                 '5396fdf3-ceb3-46b6-aaf9-454a37278bb4'])
+            >>> READ.getDataSetsResults(dstype='singleTimeSeries')
+
+        """
+
+        # make sure one of the three arguments has been sent in
+        if all(v is None for v in [ids, codes, uuids]):
+            raise ValueError('Expected DataSetID OR DataSetUUID OR DataSetCode argument')
+
+        q = self._session.query(DataSetsResults)\
+            .join(DataSets)
+        if ids:
+            q = q.filter(DataSets.DataSetID.in_(ids))
+        if codes:
+            q = q.filter(DataSets.DataSetCode.in_(codes))
+        if uuids:
+            q.filter(DataSets.DataSetUUID.in_(uuids))
+        if dstype:
+            q = q.filter(DataSets.DataSetTypeCV == dstype)
+        try:
+            return q.all()
+        except Exception as e:
+            print('Error running Query {}'.format(e))
+        return None
+
+    def getDataSetsValues(self, ids=None, codes=None, uuids=None, dstype=None):
```

Comment:
Where is the docstring for this function?

Reply:
I knew I forgot something on there! I will add that right now